### PR TITLE
add typing information

### LINKED ISSUE #274: When will be the time to drop Python 2 compatibility?
I recently attempted to use [one of my favorite static analysis tools](https://www.youtube.com/watch?v=abvW0mOrDiY) with `hpack` and it choked on
```
try:  # pragma: no cover
    basestring = basestring
except NameError:  # pragma: no cover
    basestring = (str, bytes)
```
in `hpack.py` because it doesn't support analyzing Python 2 source. My impression at this time is that this isn't a problem with the tool; that more than four years after [the Python 2 turndown](https://pythonclock.org) it's appropriate for tools to drop support for analyzing Python 2 source.

------------
So... while I see that https://hyper.readthedocs.io/en/latest advertises that the `hyper` family of projects supports `2.7.9`-and-later in the Python 2 line, do y'all `hyper` maintainers have a notion of when you're going to be dropping Python 2 compatibility from the source?

### PR DESCRIPTION
**Pull request by Kriechi**
also closes #274

### CODE REVIEW

File: src/hpack/table.py [hunk 1]
```diff
@@ -218,14 +220,14 @@ def _shrink(self):
         self._current_size = cursize
 
 
-def _build_static_table_mapping():
+def _build_static_table_mapping() -> dict[bytes, tuple[int, dict[Any, Any]]]:
```
Comment (BYK) [inline src/hpack/table.py:223]: Surprised to see `dict[Any, Any]` here. Can they really be anything or was this an escape hatch to reduce further complexity in types?

**Review by BYK — APPROVED** [review]
Shall we add `black` too in a follow up?

File: src/hpack/table.py (hunk 1 above)
Comment (Kriechi) [reply src/hpack/table.py:223]: You correctly identified the escape hatch! I wasn't yet able to derive the correct type from looking at the tests and code. Happy to change it to something more specific!

File: src/hpack/table.py [hunk 2]
```diff
@@ -1,14 +1,14 @@
...
 import logging
+from typing import Any, Optional, Union
 
 from .exceptions import InvalidTableIndex
 
 log = logging.getLogger(__name__)
 
 
-def table_entry_size(name, value):
+def table_entry_size(name: Union[bytes, str], value: Union[bytes, str]) -> int:
```
[the remark this answers is not in the document]
Comment (Kriechi) [reply src/hpack/table.py:11]: Yes! It was on my ToDo list, but didn't make it for this first PR - happy to be refactored in a future PR!
I'm working on a similar PR for the h2 library as well, so we might want to hold off on type aliases until we have a full picture.

File: src/hpack/huffman.py [hunk 3]
```diff
@@ -48,19 +47,19 @@ def encode(self, bytes_to_encode):
 
         # Convert the number to hex and strip off the leading '0x' and the
         # trailing 'L', if present.
-        final_num = hex(final_num)[2:].rstrip('L')
+        s = hex(final_num)[2:].rstrip('L')
```
[the remark this answers is not in the document]
Comment (Kriechi) [reply src/hpack/huffman.py:50]: The main reason to this change is that final_num already has a type defined, and this new assignment here would change the type, so mypy is rightfully complaining. Using a new variable name gives us a fresh type assignment. I think there were a handful of such issues that I fixed by re-naming the variable at a certain point in the function code path.

While I agree that `s` is not a perfect variable name - it is a short-lived one in a narrow scope.

File: src/hpack/table.py (hunk 1 above)
Comment (Kriechi) [reply src/hpack/table.py:223]: Finally found the correct type - will be fixed in the next push!

File: src/hpack/struct.py [hunk 4]
```diff
@@ -37,3 +39,10 @@ class NeverIndexedHeaderTuple(HeaderTuple):
     __slots__ = ()
 
     indexable = False
+
+    def __new__(cls, *args: Any) -> "NeverIndexedHeaderTuple":
+        return tuple.__new__(cls, args)
+
+
+# explicitly mentioning all valid header types , even if some superseed each other
+type Headers = Iterable[Union[HeaderTuple, NeverIndexedHeaderTuple, tuple[Union[bytes, str], Union[bytes, str]]]]
```
Comment (BYK) [inline src/hpack/struct.py:48]: Better not to allow mixed `bytes` and `str` headers as we don't expect them.

```suggestion
type Headers = Iterable[Union[HeaderTuple, NeverIndexedHeaderTuple, tuple[bytes, bytes], tuple[str, str]]]
```

Comment (Kriechi) [reply src/hpack/struct.py:48]: I had that originally but then made it a bit softer to accommodate other typing issues in the h2 project ... I'll try to make it stricter again, as suggested here.

**Pull request merged**